File: Data-Structures-and-Algorithms-using-Swift/Sorting/BucketSort.swift
import Foundation

// Allowed values from 0 to maxValue.
func bucketSort(_ arr : inout [Int], _ maxValue : Int) {
    let numBucket : Int = 5
    bucketSortUtil( &arr,maxValue,numBucket)
}

func bucketSortUtil(_ arr : inout [Int], _ maxValue : Int, _ numBucket : Int) {
    let length : Int = arr.count
    if (length == 0) {
        return
    }

    var bucket :  [ [Int]] =  [ [Int]]()
    var i : Int = 0
    // Create empty buckets
    while (i < numBucket) {
        bucket.append( [Int]())
        i += 1
    }
    
    let div : Int = Int(ceil(Double(maxValue) / Double(numBucket)))
    
    // Add elements into the buckets
    i = 0
    while (i < length) {
        if (arr[i] < 0 || arr[i] > maxValue) {
            print("Value out of range.")
            return
        }
        var bucketIndex : Int = (arr[i] / div)
        // Maximum value will be assigned to last bucket.
        if (bucketIndex >= numBucket) {
            bucketIndex = numBucket - 1
        }
        bucket[bucketIndex].append(arr[i])
        i += 1
    }
    
    // Sort the elements of each bucket.
    i = 0
    while (i < numBucket) {
        bucket[i] = bucket[i].sorted(by: <)
        i += 1
    }
    
    // Populate output from the sorted subarray.
    var index : Int = 0
    var count : Int

    i = 0
    while (i < numBucket) {
        let temp :  [Int] = bucket[i]
        count = temp.count
        var  j : Int = 0
        while (j < count) {
            arr[index] = temp[j]
            index += 1
            j += 1
        }
        i += 1
    }
}

// Testing Code.
var array : [Int] = [1, 34, 7, 99, 5, 23, 45, 88, 77, 19, 91, 100]
let maxValue : Int = 100
bucketSort( &array, maxValue)
print(array)

// [1, 5, 7, 19, 23, 34, 45, 77, 88, 91, 99, 100]

File: Data-Structures-Algorithms-In-TS/Sorting/BucketSort.ts
function bucketSort(arr: number[], maxValue: number): void {
    let numBucket = 5;
    bucketSortUtil(arr, maxValue, numBucket);
}

function bucketSortUtil(arr: number[], maxValue: number, numBucket: number): void {
    let length = arr.length;
    if (length === 0) {
        return;
    }
    let bucket: number[][] = [];

    for (let i = 0; i < numBucket; i++) {
        bucket.push([]);
    }

    let div = Math.ceil(maxValue / numBucket);

    for (let i = 0; i < length; i++) {
        if (arr[i] < 0 || arr[i] > maxValue) {
            console.log("Value out of range.");
            return;
        }

        let bucketIndex = Math.floor(arr[i] / div);

        if (bucketIndex >= numBucket)
            bucketIndex = numBucket - 1;
        bucket[bucketIndex].push(arr[i]);
    }

    for (let i = 0; i < numBucket; i++) {
        bucket[i].sort((a, b) => a - b);
    }

    let index = 0, count = 0;
    for (let i = 0; i < numBucket; i++) {
        let temp = bucket[i];
        count = temp.length;
        for (let j = 0; j < count; j++) {
            arr[index++] = temp[j];
        }
    }
}

// Testing code.
let array: number[] = [1, 34, 7, 99, 5, 23, 45, 88, 77, 19, 91, 100];
let maxValue: number = 100;
bucketSort(array, maxValue);
console.log(JSON.stringify(array));

/*
[1,5,7,19,23,34,45,77,88,91,99,100]
*/
